server/interfaces/interfaces.go

```diff
@@ -238,6 +239,74 @@ type WorkflowService interface {
 	ServeHTTP(w http.ResponseWriter, r *http.Request)
 }
 
+type GitProviders []GitProvider
+
+type GitProvider interface {
+	// MatchRepoURL returns whether a given repo URL should be handled by this
+	// provider.
+	MatchRepoURL(u *url.URL) bool
+	// MatchWebhookRequest returns whether a given webhook request was sent by
+	// this git provider.
+	MatchWebhookRequest(req *http.Request) bool
+	// ParseWebhookData parses webhook data from the given HTTP request sent to
+	// a webhook endpoint. It should only be called if MatchWebhookRequest returns
+	// true.
+	ParseWebhookData(req *http.Request) (*WebhookData, error)
+	// IsRepoPrivate returns whether the repo is only viewable by its owner and
+	// trusted users.
+	IsRepoPrivate(ctx context.Context, accessToken, repoURL string) (bool, error)
+	// RegisterWebhook registers the given webhook URL to listen for push and
+	// pull request (also called "merge request") events.
+	RegisterWebhook(ctx context.Context, accessToken, repoURL, webhookURL string) (string, error)
```

Comment:
FYI the returned webhook URL is a string even though GitHub returns int64. Some Git providers use UUIDs so we'll need a string to represent that. The plan is to remove the GitHubWebhookID from the DB and migrate to a generic GitProviderWebhookID string column. (I'll just manually re-create our workflows in the UI so we don't have to bother with a DB migration of GitHubWebhookID -> GitProviderWebhookID)

Reply:
ok, SGTM